test/e2e/logging_test.go

```diff
@@ -0,0 +1,188 @@
+		test.TearDown(clients, names)
+	}
+	test.CleanupOnInterrupt(cleanup)
+	t.Cleanup(cleanup)
+
+	t.Log("Creating a new Service")
+
+	resources, err := v1test.CreateServiceReady(t, clients, &names, []rtesting.ServiceOption{
+		rtesting.WithConfigAnnotations(map[string]string{
+			autoscaling.MinScaleAnnotationKey: "1",
+			autoscaling.MaxScaleAnnotationKey: "1",
+		})}...)
+	if err != nil {
+		t.Fatalf("Failed to create initial Service: %v: %v", names.Service, err)
+	}
+
+	_, err = sendRequest(t, clients, test.ServingFlags.ResolvableDomain, resources.Route.Status.URL.URL())
+	if err != nil {
+		t.Fatal("Unexpected error when sending request to helloworld:", err)
+	}
+
+	pod, err := theOnlyPod(clients, resources.Revision.Namespace, resources.Revision.Name)
+	if err != nil {
+		t.Fatalf("Fail to fetch the pod: %v", err)
+	}
+
+	// A request was sent to / in WaitForEndpointState.
+	if err := waitForLog(t, clients, pod.Namespace, pod.Name, "queue-proxy", func(log map[string]interface{}) bool {
+		v, ok := log["requestUrl"]
+		if !ok {
+			return false
+		}
+		if v.(string) != "/" {
+			return false
+		}
+
+		v, ok = log["userAgent"]
+		if !ok {
+			return false
+		}
+		return v.(string) != network.QueueProxyUserAgent
+	}); err != nil {
+		t.Fatalf("Got error waiting for normal request logs: %v", err)
+	}
+
+	// Health check requests are sent to / with a specific userAgent value periodically.
+	if err := waitForLog(t, clients, pod.Namespace, pod.Name, "queue-proxy", func(log map[string]interface{}) bool {
+		v, ok := log["requestUrl"]
+		if !ok {
+			return false
+		}
+		if v.(string) != "/" {
+			return false
+		}
+
+		v, ok = log["userAgent"]
+		if !ok {
+			return false
+		}
+		return v.(string) == network.QueueProxyUserAgent
+
+	}); err != nil {
+		t.Fatalf("Got error waiting for health check log: %v", err)
+	}
+}
+
+func theOnlyPod(clients *test.Clients, ns, rev string) (corev1.Pod, error) {
+	pods, err := clients.KubeClient.Kube.CoreV1().Pods(ns).List(metav1.ListOptions{LabelSelector: fmt.Sprintf("app=%s", rev)})
+
+	if err != nil {
+		return corev1.Pod{}, err
+	}
+
+	if len(pods.Items) == 0 {
+		return corev1.Pod{}, fmt.Errorf("Expect 1 pod, but got %d for %s:%s", len(pods.Items), system.Namespace(), rev)
+	}
+
+	return pods.Items[0], nil
+}
+
+// waitForLog fetches the logs from a container of a pod decided by the given parameters
+// until the given condition is meet or timeout. Most of knative logs are in json format.
+func waitForLog(t *testing.T, clients *test.Clients, ns, podName, container string, condition func(log map[string]interface{}) bool) error {
+	return wait.PollImmediate(time.Second, 30*time.Second, func() (bool, error) {
+		podLogOpts := corev1.PodLogOptions{Container: container}
+
+		req := clients.KubeClient.Kube.CoreV1().Pods(ns).GetLogs(podName, &podLogOpts)
+		podLogs, err := req.Stream()
+		if err != nil {
+			return false, err
+		}
+		defer podLogs.Close()
+
+		buf := new(bytes.Buffer)
+		_, err = io.Copy(buf, podLogs)
```

Comment:
We should also remove the buf outside of the loop and just `Reset` it on every loop (it would still retain the storage).

Reply:
Done